docs/configuration.md
# Configuration

The Pacifica Core services require two configuration files. The REST
API utilizes [CherryPy](https://github.com/cherrypy) and review of
their
[configuration documentation](http://docs.cherrypy.org/en/latest/config.html)
is recommended. The service configuration file is a INI formatted
file containing configuration for database connections.

## CherryPy Configuration File

An example of Policy server CherryPy configuration:

```ini
[global]
log.screen: True
log.access_file: 'access.log'
log.error_file: 'error.log'
server.socket_host: '0.0.0.0'
server.socket_port: 8181

[/]
request.dispatch: cherrypy.dispatch.MethodDispatcher()
tools.response_headers.on: True
tools.response_headers.headers: [('Content-Type', 'application/json')]
```

## Service Configuration File

The service configuration is an INI file and an example is as follows:

```ini
[policy]
; This section has policy service specific config options

; The following strings reference formatting directives {}. The
; object passed to the format method is the transaction object
; from the metadata API. The DOI is special and added into the
; transaction object for that format as well.

; Internal URL format for transactions not released or have DOIs
internal_url_format = https://internal.example.com/{_id}

; Release URL format for transactions released but no DOI
release_url_format = https://release.example.com/{_id}

; DOI URL format for transactions with a DOI
doi_url_format = https://dx.doi.org/{doi}

; In memory object cache size (used in data release)
cache_size = 10000

; This sets the admin group name
admin_group = admin

; This sets the admin group id (should match group name in metadata)
admin_group_id = 0

; This sets the admin user id (should match user name in metadata)
admin_user_id = 0

[metadata]
; This section contains configuration for metadata service

; The global metadata url
endpoint_url = http://localhost:8121

; The endpoint to check for status of metadata service
status_url = http://localhost:8121/groups

```

## Starting the Service

Starting the Policy service can be done by two methods. However,
understanding the requirements and how they apply to REST services
is important to address as well. Using the
internal CherryPy server to start the service is recommended for
Windows platforms. For Linux/Mac platforms it is recommended to
deploy the service with
[uWSGI](https://uwsgi-docs.readthedocs.io/en/latest/).

### Deployment Considerations

The Policy server can have the same memory consumption issues as the
Metadata service. Please consider those recommendations
[here](https://pacifica-metadata.readthedocs.io/en/latest/configuration.html#deployment-considerations)
similarly for the Policy service.

### CherryPy Server

To make running the Policy service using the CherryPy's builtin
server easier we have a command line entry point.

```
$ pacifica-policy --help
usage: pacifica-policy [-h] [-c CONFIG] [-p PORT] [-a ADDRESS]

Run the policy server.

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG, --config CONFIG
                        cherrypy config file
  -p PORT, --port PORT  port to listen on
  -a ADDRESS, --address ADDRESS
                        address to listen on
$ pacifica-policy
[09/Jan/2019:09:17:26] ENGINE Listening for SIGTERM.
[09/Jan/2019:09:17:26] ENGINE Bus STARTING
[09/Jan/2019:09:17:26] ENGINE Set handler for console events.
[09/Jan/2019:09:17:26] ENGINE Started monitor thread 'Autoreloader'.
[09/Jan/2019:09:17:26] ENGINE Serving on http://0.0.0.0:8181
[09/Jan/2019:09:17:26] ENGINE Bus STARTED
```

### uWSGI Server

To make running the Policy service using uWSGI easier we have a
module to be included as part of the uWSGI configuration. uWSGI is
very configurable and can use this module many different ways. Please
consult the
[uWSGI Configuration](https://uwsgi-docs.readthedocs.io/en/latest/Configuration.html)
documentation for more complicated deployments.

```
$ pip install uwsgi
$ uwsgi --http-socket :8181 --master --module pacifica.policy.wsgi
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- Endpoints for status
- Endpoints for reporting

## [0.8.3] - 2020-05-11
### Updated
- Pull #118 Remove Requirements pip>=20 by [@dmlb2000](https://github.com/dmlb2000)
- Pull #119 Remove Elasticsearch and Six by [@dmlb2000](https://github.com/dmlb2000)

## [0.8.2] - 2020-01-22
### Updated
- Pull #116 Add Python 3.7 and 3.8 by [@dmlb2000](https://github.com/dmlb2000)
- Pull #115 Update Pylint 2.0+ by [@dmlb2000](https://github.com/dmlb2000)
- Pull #114 Remove Python 2.7 support by [@dmlb2000](https://github.com/dmlb2000)

## [0.8.1] - 2019-11-08
### Updated
- Pull #112 Allow admins to see all events by [@dmlb2000](https://github.com/dmlb2000)

## [0.8.0] - 2019-09-03
### Removed
- Deprecated SearchSync (closed #107 and #95) functionality by [@dmlb2000](https://github.com/dmlb2000)
- Pull #110 Refactored testing suite by [@dmlb2000](https://github.com/dmlb2000)

## [0.7.0] - 2019-06-21
### Changed
- Fix #100 add better exclude command line by [@dmlb2000](https://github.com/dmlb2000)
- Fix #102 add config file caching by [@dmlb2000](https://github.com/dmlb2000)
- Fix #99 add searchsync performance testing by [@dmlb2000](https://github.com/dmlb2000)

## [0.6.7] - 2019-06-21
### Changed
- Fix #93 add exclude searchsync options by [@dmlb2000](https://github.com/dmlb2000)
- Fix #91 add better docs by [@dmlb2000](https://github.com/dmlb2000)

## [0.6.6] - 2019-05-30
### Changed
- Add field data flag to the has doi metadata attribute by [@plithnar](https://github.com/plithnar)

## [0.6.5] - 2019-05-23
### Changed
- Fix #84 make the elasticsearch endpoint not constant
- Fix #85 verify event policy is of type ingest

## [0.6.4] - 2019-05-22
### Added
- Allowed for Elasticsearch sniffing to be configurable (Pull #78)

### Changed
- Fix #79 Change Web Root to be Status
- Fix #81 Modify Elasticsearch Mappings to include more fields

## [0.6.1] - 2019-05-18
### Added
- Events policy endpoint
- Ingest policy endpoint
- Uploader policy endpoint
- ElasticSearch synchronization
- Data release management

### Changed


pacifica/policy/config.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
"""Configuration reading and validation module."""
from os import getenv
from configparser import ConfigParser as SafeConfigParser
from functools import lru_cache
from pacifica.policy.globals import CONFIG_FILE


@lru_cache(maxsize=1)
def get_config():
    """
    Return the ConfigParser object with defaults set.

    Currently metadata API doesn't work with SQLite the queries are
    too complex and it only is supported with MySQL and PostgreSQL.
    """
    configparser = SafeConfigParser()
    configparser.add_section('policy')
    configparser.set(
        'policy', 'internal_url_format',
        getenv('INTERNAL_URL_FORMAT',
               'https://internal.example.com/{_id}')
    )
    configparser.set(
        'policy', 'release_url_format',
        getenv('RELEASE_URL_FORMAT',
               'https://release.example.com/{_id}')
    )
    configparser.set(
        'policy', 'doi_url_format',
        getenv('DOI_URL_FORMAT', 'https://dx.doi.org/{doi}')
    )
    configparser.set(
        'policy', 'cache_size',
        getenv(
            'CACHE_SIZE',
            '10000'
        )
    )
    configparser.set(
        'policy', 'admin_group',
        getenv(
            'ADMIN_GROUP',
            'admin'
        )
    )
    configparser.set(
        'policy', 'admin_group_id',
        getenv('ADMIN_GROUP_ID', '0')
    )
    configparser.set(
        'policy', 'admin_user_id',
        getenv('ADMIN_USER_ID', '0')
    )
    configparser.add_section('metadata')
    configparser.set(
        'metadata', 'endpoint_url',
        getenv(
            'METADATA_URL',
            'http://localhost:8121'
        )
    )
    configparser.set(
        'metadata', 'status_url',
        getenv(
            'STATUS_URL',
            'http://localhost:8121/groups'
        )
    )
    configparser.read(CONFIG_FILE)
    return configparser
